src/lang-promql/parser/parser.ts

```diff
@@ -0,0 +1,298 @@
+    case Expr:
+      return getType(node.firstChild);
+    case AggregateExpr:
+      return NodeTypeVector;
+    case VectorSelector:
+      return NodeTypeVector;
+    case StringLiteral:
+      return NodeTypeString;
+    case NumberLiteral:
+      return NodeTypeScalar;
+    case MatrixSelector:
+      return NodeTypeMatrix;
+    case SubqueryExpr:
+      return NodeTypeMatrix;
+    case ParenExpr:
+      return getType(walkThrough(node, Expr));
+    case UnaryExpr:
+      return getType(walkThrough(node, Expr));
+    case BinaryExpr:
+      const lt = getType(node.firstChild);
+      const rt = getType(node.lastChild);
+      if (lt === NodeTypeScalar && rt === NodeTypeScalar) {
+        return NodeTypeScalar;
+      }
+      // TODO what happen if you have a stringLiteral instead
+      return NodeTypeVector;
+    case FunctionCall:
+      const funcNode = node.firstChild?.firstChild;
+      if (!funcNode) {
+        return NodeTypeNone;
+      }
+      if (funcNode.type.id === Time || funcNode.type.id === Scalar) {
+        return NodeTypeScalar;
+      }
+      return NodeTypeVector;
+    default:
+      return NodeTypeNone;
+  }
+}
+
+export class Parser {
+  private readonly tree: Tree;
+  private readonly diagnostics: Diagnostic[];
+
+  constructor(tree: Tree) {
+    this.tree = tree;
+    this.diagnostics = [];
+  }
+
+  getDiagnostics(): Diagnostic[] {
+    return this.diagnostics.sort((a, b) => {
+      return a.from - b.from;
+    });
+  }
+
+  analyze() {
+    this.checkAST(this.tree);
+    this.diagnosticAllErrorNode();
+  }
+
+  private diagnosticAllErrorNode() {
+    this.tree.iterate({
+      enter: (type, start, end) => {
+        // usually there is an error node at the end of the expression when user is typing
+        // so it's not really a useful information to say the expression is wrong.
+        // Hopefully if there is an error node at the end of the tree, checkAST should yell more precisely
+        if (type.id === 0 && end !== this.tree.end) {
+          // get the parent to highlight the expression where the errorNode has been placed
+          const node = this.tree.resolve(start, -1);
+          this.diagnostics.push({
+            severity: 'error',
+            message: 'unexpected expression',
+            from: node.start,
+            to: node.end,
+          });
+        }
+      },
+    });
+  }
+
+  // checkAST is inspired of the same named method from prometheus/prometheus:
+  // https://github.com/prometheus/prometheus/blob/master/promql/parser/parse.go#L433
+  private checkAST(node: Subtree | undefined | null): NodeType {
+    if (!node) {
+      return NodeTypeNone;
+    }
+    switch (node.type.id) {
+      case Expr:
+        return this.checkAST(node.firstChild);
+      case AggregateExpr:
+        this.checkAggregationExpr(node);
+        break;
+      case BinaryExpr:
+        this.checkBinaryExpr(node);
+        break;
+      // TODO add FunctionCall management
+      case FunctionCall:
```

Comment:
Just a note:

In PromLens I have Go code that generates a JS list of all PromQL function signatures like this: https://gist.github.com/juliusv/bf342be2b1b4a34d99ab35aac0d35607

Could be nice to put this into a central place sometime and use that from here and PromLens...

(I also have a similar thing for function docs)

Reply:
ah nice you generated it. On my side it was copy and past and translate it to Typescript ^^.

Reply:
since I ended up to have it here, I could export it so you can use it. ( For the moment I didn't export it)

Reply:
and yeah having it in a central place will make sense since we are both using it

Reply:
At some point it probably makes sense to have a Typescript PromQL repo / package that just has all the metadata like function names + signatures, lists of token strings, etc. etc. - but ok for now :)

Reply:
I would say yes if one day there is a need to have access to this data without using codemirror-promql. But maybe that's already what you are doing in PromLabs